### PR TITLE
Add `has-background` classes to pullquote and Media & Text blocks

### LINKED ISSUE #11172: Should class has-background be in blocks where you can add background color
At the moment there is  class `has-background` in paragraph block. But I don't see it in other blocks where you can add background color. For example in Media Text block.

We could target something like `[class$="background-color"]`. But I'd find more helpful to be consistent on outputted class names. 

Suggestion: Every block that supports bg color should have `has-background` class.

### PR DESCRIPTION
**Pull request by chrisvanpatten**
Fixes #11172.

This adds `has-background` classes to the pullquote and Media & Text blocks which makes them more consistent with the paragraph block.

I also did some light refactoring to try to make the code easier to read (which did make it easier to write the PR) and made it so neither block has an inline `background-color` style if using a named color value.

There should be no visual changes in the editor or front-end.

I expect the tests on this are going to fail because I didn't update any snapshots; I plan to fix that up in the morning.

### CODE REVIEW

**Comment by chrisvanpatten** [comment]
Hmm… weirdly, the e2e tests pass but now the unit tests fail... I will take another look in the morning but if anyone sees an obvious mistake please feel free to push to this branch 👍 For now, bed beckons. 🛌

**Comment by chrisvanpatten** [comment]
Tests should pass now. Got tripped up the `classnames` doesn't return `undefined`, just an empty string, when there are no classes to apply. Super irritating (and weird that I didn't notice this until today).

In any event, ready for a review ❤️

**Comment by chrisvanpatten** [comment]
screenshot 2018-10-30 at 08 33 59

Not clear why Travis failed, because it says the tests passed…

Going to restart the build and see what happens…

**Comment by youknowriad** [comment]
@chrisvanpatten That's because master was broken. It should be fixed now, if the tests don't pass, try to rebase.

**Comment by chrisvanpatten** [comment]
Thanks for the tip @youknowriad — tests are now passing!

**Review by aduth — CHANGES_REQUESTED** [review]
In brief testing, this invalidates a Media & Text block which had been previously saved with a background color, when loaded in this branch. I'd think this shouldn't be happening generally, as maybe we should consider _additional_ classes as not-invalidating. Without such a solution or a block deprecation, I don't think this should go through in its current form.

**Comment by chrisvanpatten** [comment]
> In brief testing, this invalidates a Media & Text block which had been previously saved with a background color, when loaded in this branch. I'd think this shouldn't be happening generally, as maybe we should consider additional classes as not-invalidating. Without such a solution or a block deprecation, I don't think this should go through in its current form.

That's totally fair. I agree that we shouldn't invalidate blocks by classNames changing but unless there's a mechanism in Gutenberg core I don't think it's worth pushing this in.

(It could definitely be fixed with a deprecation, but that feels like using an impact driver when you really need a jeweler's screwdriver.)

**Comment by aduth** [comment]
@samikeijonen Correct, at least in the current validation procedure. As mentioned in my previous comment, this is probably too heavy-handed. The validator errs on the side of caution when it comes to semantic differences in markup, of which even an added class could fall under (I can imagine a scenario where an added class has a "destructive" effect). 

Generally speaking, _adding_ a class could probably be exempted from the validation behavior, or at least accounted for by the behavior of `customClassName`.

**Comment by aduth** [comment]
On second thought, while all of the above is still true, given the way a deprecation works, it's something that if we were to add one now, at the point we resolve the framework-level issue of handling additional classes, we could easily drop the deprecation without concern.

Is adding a deprecation something you'd be able to do here?

**Comment by samikeijonen** [comment]
> Is adding a deprecation something you'd be able to do here?

That was question for @chrisvanpatten I assume?

> at the point we resolve the framework-level issue of handling additional classes, we could easily drop the deprecation without concern.

That would be wonderful if at some point we could add classes without deprecation logic.

**Comment by chrisvanpatten** [comment]
Right now I don’t have the bandwidth to take this further but I’m more than happy to let someone take over the PR!

**Comment by gziolo** [comment]
@jorgefilipecosta would you mind taking it over?

**Comment by jorgefilipecosta** [comment]
Sure I will catch with the discussions and try to take the PR to the finish line. Thank you to all those involved.

**Comment by chrisvanpatten** [comment]
Thanks @gziolo @jorgefilipecosta 🎉 ❤️

File: packages/block-library/src/media-text/edit.js [hunk 1]
```diff
@@ -117,12 +117,13 @@ class MediaTextEdit extends Component {
...
 			'has-media-on-the-right': 'right' === mediaPosition,
 			'is-selected': isSelected,
+			'has-background': ( backgroundColor.class || backgroundColor.color ),
 			[ backgroundColor.class ]: backgroundColor.class,
 		} );
 		const widthString = `${ temporaryMediaWidth || mediaWidth }%`;
 		const style = {
 			gridTemplateColumns: 'right' === mediaPosition ? `auto ${ widthString }` : `${ widthString } auto`,
-			backgroundColor: backgroundColor.color,
+			backgroundColor: backgroundColor.class ? undefined : backgroundColor.color,
```
Comment (jorgefilipecosta) [inline packages/block-library/src/media-text/edit.js]: It was decided to use inline color styles in the editor so even if the theme developer is not loading the editor styles with the color classes, the color is still viewable in the editor. So I will revert this line change during the update.

**Comment by jorgefilipecosta** [comment]
This PR is updated and ready for a review.

**Comment by aduth** [comment]
Could you rebase this one? It has the `gutenberg-mobile` branch from eons ago, so I'd have a hard time guaranteeing it still works as-is.

I'm still a bit reluctant on the deprecations, still all previous comments held true, I have doubts we'd know to drop those deprecations once we improve the validation. I'd like to set some time aside looking at the validation, though still not want to let this one be blocked. I'll see what I can do there this week.

**Comment by mkronenfeld** [comment]
@jorgefilipecosta Any progress since February? This little class attribute has a huge impact for page layouts, since every other layout block with background color has it, too 😅

**Comment by jorgefilipecosta** [comment]
Hi @chrisvanpatten, I tried to update this PR but when I tried to push to your fork I got the following error:
```
Pushing to https://github.com/TomodomoCo/gutenberg/
remote: Permission to TomodomoCo/gutenberg.git denied to jorgefilipecosta.
fatal: unable to access 'https://github.com/TomodomoCo/gutenberg/': The requested URL returned error: 403
Completed with errors, see above
```
It seems the option to  Allow edits from maintainers was disabled.


I created another PR so I could continue these changes https://github.com/WordPress/gutenberg/pull/17529.

**Pull request closed without merging**